File: scripts/firebase-deploy-standalone.js
#!/usr/bin/env node

/**
 * Standalone Firebase Hosting Deployment Script
 * 
 * This script deploys static websites to Firebase Hosting using only the REST API.
 * No Firebase CLI required!
 * 
 * Requirements:
 * - Node.js 14+
 * - Valid Google OAuth access token with Firebase permissions
 * 
 * Usage:
 *   node firebase-deploy-standalone.js --project=my-project --token=ya29.xxx --dir=./public
 * 
 * Environment Variables (alternative to CLI args):
 *   FIREBASE_PROJECT_ID - Your Firebase project ID
 *   FIREBASE_ACCESS_TOKEN - Your Google OAuth access token
 *   DEPLOY_DIR - Directory containing files to deploy (default: ./public)
 *   DEPLOY_CHANNEL - Deployment channel (default: live, can be preview or custom)
 */

const fs = require('fs').promises;
const path = require('path');
const https = require('https');

// Configuration
const config = {
  projectId: process.env.FIREBASE_PROJECT_ID || getArg('--project'),
  accessToken: process.env.FIREBASE_ACCESS_TOKEN || getArg('--token'),
  deployDir: process.env.DEPLOY_DIR || getArg('--dir') || './public',
  channel: process.env.DEPLOY_CHANNEL || getArg('--channel') || 'live',
};

// Helper to get CLI arguments
function getArg(name) {
  const arg = process.argv.find(a => a.startsWith(name + '='));
  return arg ? arg.split('=')[1] : null;
}

// Helper to make HTTPS requests
function httpsRequest(url, options = {}) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const reqOptions = {
      hostname: urlObj.hostname,
      path: urlObj.pathname + urlObj.search,
      method: options.method || 'GET',
      headers: {
        'Authorization': `Bearer ${config.accessToken}`,
        ...(options.headers || {}),
      },
    };

    const req = https.request(reqOptions, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          try {
            resolve({ status: res.statusCode, data: JSON.parse(data) });
          } catch (e) {
            resolve({ status: res.statusCode, data: data });
          }
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${data}`));
        }
      });
    });

    req.on('error', reject);

    if (options.body) {
      req.write(typeof options.body === 'string' ? options.body : JSON.stringify(options.body));
    }

    req.end();
  });
}

// Check if Firebase project exists
async function checkFirebaseProject() {
  console.log(`🔍 Checking if Firebase project exists: ${config.projectId}`);
  try {
    await httpsRequest(`https://firebase.googleapis.com/v1beta1/projects/${config.projectId}`);
    console.log('✅ Project exists');
    return true;
  } catch (error) {
    console.error('❌ Project does not exist or you don\'t have access');
    console.error('\nPlease ensure:');
    console.error(`1. Firebase project "${config.projectId}" exists`);
    console.error('2. You have the necessary permissions');
    console.error('3. Your access token is valid');
    console.error('\nCreate project at: https://console.firebase.google.com/');
    return false;
  }
}

// Check if Hosting is initialized
async function checkHostingInitialized() {
  const siteId = config.projectId;
  console.log(`🔍 Checking if Hosting is initialized for site: ${siteId}`);
  
  try {
    await httpsRequest(
      `https://firebasehosting.googleapis.com/v1beta1/projects/${config.projectId}/sites/${siteId}`
    );
    console.log('✅ Hosting is initialized');
    return true;
  } catch (error) {
    console.error('❌ Hosting is not initialized');
    console.error('\nPlease initialize Hosting:');
    console.error(`1. Go to https://console.firebase.google.com/project/${config.projectId}/hosting`);
    console.error('2. Click "Get started" to initialize Hosting');
    console.error('3. Return here and run the script again');
    return false;
  }
}

// Read all files from directory recursively
async function readDirectory(dir, baseDir = dir) {
  const files = [];
  const items = await fs.readdir(dir, { withFileTypes: true });

  for (const item of items) {
    const fullPath = path.join(dir, item.name);
    
    if (item.isDirectory()) {
      // Recursively read subdirectories
      files.push(...await readDirectory(fullPath, baseDir));
    } else if (item.isFile()) {
      try {
        // Read file with error handling
        const content = await fs.readFile(fullPath, 'utf-8');
        const relativePath = '/' + path.relative(baseDir, fullPath).replace(/\\/g, '/');
        
        // Validate file size (10MB limit per file)
        const sizeInMB = Buffer.byteLength(content, 'utf-8') / 1024 / 1024;
        if (sizeInMB > 10) {
          console.warn(`⚠️  Warning: ${relativePath} is ${sizeInMB.toFixed(2)}MB (limit: 10MB), skipping...`);
          continue;
        }
        
        files.push({ path: relativePath, content });
      } catch (error) {
        console.warn(`⚠️  Warning: Failed to read ${fullPath}: ${error.message}`);
        // Continue with other files
      }
    }
  }

  return files;
}

// Create a new Hosting version
async function createHostingVersion() {
  const siteId = config.projectId;
  console.log('📝 Creating new hosting version...');

  const response = await httpsRequest(
    `https://firebasehosting.googleapis.com/v1beta1/projects/${config.projectId}/sites/${siteId}/versions`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: {
        config: {
          headers: [{
            glob: '**',
            headers: {
              'Cache-Control': 'public, max-age=3600',
            },
          }],
        },
      },
    }
  );

  console.log(`✅ Version created: ${response.data.name}`);
  return response.data.name;
}

// Upload files to the version
async function uploadFiles(versionName, files) {
  console.log(`📤 Uploading ${files.length} file(s)...`);

  // Convert files to base64
  const fileList = {};
  for (const file of files) {
    fileList[file.path] = Buffer.from(file.content, 'utf-8').toString('base64');
    console.log(`   - ${file.path} (${(file.content.length / 1024).toFixed(2)} KB)`);
  }

  await httpsRequest(
    `https://firebasehosting.googleapis.com/v1beta1/${versionName}:populateFiles`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: { files: fileList },
    }
  );

  console.log('✅ Files uploaded successfully');
}

// Finalize the version
async function finalizeVersion(versionName) {
  console.log('🔨 Finalizing version...');

  await httpsRequest(
    `https://firebasehosting.googleapis.com/v1beta1/${versionName}?update_mask=status`,
    {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: { status: 'FINALIZED' },
    }
  );

  console.log('✅ Version finalized');
}

// Create a release
async function createRelease(versionName) {
  const siteId = config.projectId;
  console.log(`🚀 Creating release on channel: ${config.channel}...`);

  const releaseUrl = config.channel === 'live'
    ? `https://firebasehosting.googleapis.com/v1beta1/projects/${config.projectId}/sites/${siteId}/releases?versionName=${versionName}`
    : `https://firebasehosting.googleapis.com/v1beta1/projects/${config.projectId}/sites/${siteId}/channels/${config.channel}/releases?versionName=${versionName}`;

  const response = await httpsRequest(releaseUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: { message: 'Deployed via standalone script' },
  });

  console.log(`✅ Release created: ${response.data.name}`);
  return response.data.name;
}

// Main deployment function
async function deploy() {
  console.log('\n🔥 Firebase Hosting Deployment Tool\n');

  // Validate configuration
  if (!config.projectId) {
    console.error('❌ Error: Firebase project ID is required');
    console.error('Provide via --project=my-project or FIREBASE_PROJECT_ID env var');
    process.exit(1);
  }

  if (!config.accessToken) {
    console.error('❌ Error: Access token is required');
    console.error('Provide via --token=ya29.xxx or FIREBASE_ACCESS_TOKEN env var');
    console.error('\nTo get an access token:');
    console.error('1. Go to OAuth 2.0 Playground: https://developers.google.com/oauthplayground/');
    console.error('2. Select Firebase Hosting API v1beta1 scopes');
    console.error('3. Authorize and get access token');
    process.exit(1);
  }

  try {
    // Step 1: Check project exists
    const projectExists = await checkFirebaseProject();
    if (!projectExists) {
      process.exit(1);
    }

    // Step 2: Check hosting initialized
    const hostingInitialized = await checkHostingInitialized();
    if (!hostingInitialized) {
      process.exit(1);
    }

    // Step 3: Read files
    console.log(`\n📂 Reading files from: ${config.deployDir}`);
    const files = await readDirectory(config.deployDir);
    console.log(`✅ Found ${files.length} file(s)`);

    if (files.length === 0) {
      console.error('❌ Error: No files found to deploy');
      process.exit(1);
    }

    // Step 4: Create version
    const versionName = await createHostingVersion();

    // Step 5: Upload files
    await uploadFiles(versionName, files);

    // Step 6: Finalize version
    await finalizeVersion(versionName);

    // Step 7: Create release
    await createRelease(versionName);

    // Success!
    const siteUrl = config.channel === 'live'
      ? `https://${config.projectId}.web.app`
      : `https://${config.projectId}--${config.channel}.web.app`;

    console.log('\n🎉 Deployment successful!');
    console.log(`\n🌐 Your site is live at: ${siteUrl}\n`);

  } catch (error) {
    console.error('\n❌ Deployment failed:', error.message);
    process.exit(1);
  }
}

// Run deployment
if (require.main === module) {
  deploy();
}

module.exports = { deploy };

File: scripts/firebase-deploy-standalone.py
#!/usr/bin/env python3

"""
Standalone Firebase Hosting Deployment Script (Python)

This script deploys static websites to Firebase Hosting using only the REST API.
No Firebase CLI required!

Requirements:
- Python 3.7+
- requests library (pip install requests)
- Valid Google OAuth access token with Firebase permissions

Usage:
    python3 firebase-deploy-standalone.py --project=my-project --token=ya29.xxx --dir=./public

Environment Variables (alternative to CLI args):
    FIREBASE_PROJECT_ID - Your Firebase project ID
    FIREBASE_ACCESS_TOKEN - Your Google OAuth access token
    DEPLOY_DIR - Directory containing files to deploy (default: ./public)
    DEPLOY_CHANNEL - Deployment channel (default: live, can be preview or custom)
"""

import os
import sys
import argparse
import json
import base64
from pathlib import Path
from typing import List, Dict, Tuple

try:
    import requests
except ImportError:
    print("❌ Error: requests library is required")
    print("Install it with: pip install requests")
    sys.exit(1)


class FirebaseDeployer:
    """Firebase Hosting deployment using REST API"""

    def __init__(self, project_id: str, access_token: str, deploy_dir: str, channel: str = "live"):
        self.project_id = project_id
        self.access_token = access_token
        self.deploy_dir = Path(deploy_dir)
        self.channel = channel
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {access_token}",
            "Content-Type": "application/json"
        })

    def check_project_exists(self) -> bool:
        """Check if Firebase project exists"""
        print(f"🔍 Checking if Firebase project exists: {self.project_id}")
        
        url = f"https://firebase.googleapis.com/v1beta1/projects/{self.project_id}"
        response = self.session.get(url)
        
        if response.status_code == 200:
            print("✅ Project exists")
            return True
        else:
            print("❌ Project does not exist or you don't have access")
            print("\nPlease ensure:")
            print(f"1. Firebase project '{self.project_id}' exists")
            print("2. You have the necessary permissions")
            print("3. Your access token is valid")
            print("\nCreate project at: https://console.firebase.google.com/")
            return False

    def check_hosting_initialized(self) -> bool:
        """Check if Hosting is initialized for the site"""
        site_id = self.project_id
        print(f"🔍 Checking if Hosting is initialized for site: {site_id}")
        
        url = f"https://firebasehosting.googleapis.com/v1beta1/projects/{self.project_id}/sites/{site_id}"
        response = self.session.get(url)
        
        if response.status_code == 200:
            print("✅ Hosting is initialized")
            return True
        else:
            print("❌ Hosting is not initialized")
            print("\nPlease initialize Hosting:")
            print(f"1. Go to https://console.firebase.google.com/project/{self.project_id}/hosting")
            print("2. Click 'Get started' to initialize Hosting")
            print("3. Return here and run the script again")
            return False

    def read_files(self) -> List[Dict[str, str]]:
        """Read all files from deploy directory"""
        print(f"\n📂 Reading files from: {self.deploy_dir}")
        
        if not self.deploy_dir.exists():
            raise FileNotFoundError(f"Deploy directory not found: {self.deploy_dir}")
        
        files = []
        for file_path in self.deploy_dir.rglob("*"):
            if file_path.is_file():
                # Read file content
                with open(file_path, "r", encoding="utf-8", errors="ignore") as f:
                    content = f.read()
                
                # Get relative path with leading slash
                relative_path = "/" + str(file_path.relative_to(self.deploy_dir)).replace("\\", "/")
                
                files.append({
                    "path": relative_path,
                    "content": content
                })
        
        print(f"✅ Found {len(files)} file(s)")
        return files

    def create_hosting_version(self) -> str:
        """Create a new hosting version"""
        site_id = self.project_id
        print("📝 Creating new hosting version...")
        
        url = f"https://firebasehosting.googleapis.com/v1beta1/projects/{self.project_id}/sites/{site_id}/versions"
        payload = {
            "config": {
                "headers": [{
                    "glob": "**",
                    "headers": {
                        "Cache-Control": "public, max-age=3600"
                    }
                }]
            }
        }
        
        response = self.session.post(url, json=payload)
        response.raise_for_status()
        
        version_data = response.json()
        version_name = version_data["name"]
        print(f"✅ Version created: {version_name}")
        return version_name

    def upload_files(self, version_name: str, files: List[Dict[str, str]]) -> None:
        """Upload files to the version"""
        print(f"📤 Uploading {len(files)} file(s)...")
        
        # Convert files to base64
        file_list = {}
        for file in files:
            # Encode content to base64
            content_bytes = file["content"].encode("utf-8")
            base64_content = base64.b64encode(content_bytes).decode("utf-8")
            file_list[file["path"]] = base64_content
            
            size_kb = len(file["content"]) / 1024
            print(f"   - {file['path']} ({size_kb:.2f} KB)")
        
        url = f"https://firebasehosting.googleapis.com/v1beta1/{version_name}:populateFiles"
        payload = {"files": file_list}
        
        response = self.session.post(url, json=payload)
        response.raise_for_status()
        
        print("✅ Files uploaded successfully")

    def finalize_version(self, version_name: str) -> None:
        """Finalize the version"""
        print("🔨 Finalizing version...")
        
        url = f"https://firebasehosting.googleapis.com/v1beta1/{version_name}?update_mask=status"
        payload = {"status": "FINALIZED"}
        
        response = self.session.patch(url, json=payload)
        response.raise_for_status()
        
        print("✅ Version finalized")

    def create_release(self, version_name: str) -> str:
        """Create a release to deploy the version"""
        site_id = self.project_id
        print(f"🚀 Creating release on channel: {self.channel}...")
        
        if self.channel == "live":
            url = f"https://firebasehosting.googleapis.com/v1beta1/projects/{self.project_id}/sites/{site_id}/releases?versionName={version_name}"
        else:
            url = f"https://firebasehosting.googleapis.com/v1beta1/projects/{self.project_id}/sites/{site_id}/channels/{self.channel}/releases?versionName={version_name}"
        
        payload = {"message": "Deployed via Python standalone script"}
        
        response = self.session.post(url, json=payload)
        response.raise_for_status()
        
        release_data = response.json()
        release_name = release_data["name"]
        print(f"✅ Release created: {release_name}")
        return release_name

    def deploy(self) -> str:
        """Main deployment workflow"""
        print("\n🔥 Firebase Hosting Deployment Tool (Python)\n")
        
        # Step 1: Check project exists
        if not self.check_project_exists():
            raise Exception("Project check failed")
        
        # Step 2: Check hosting initialized
        if not self.check_hosting_initialized():
            raise Exception("Hosting not initialized")
        
        # Step 3: Read files
        files = self.read_files()
        if not files:
            raise Exception("No files found to deploy")
        
        # Step 4: Create version
        version_name = self.create_hosting_version()
        
        # Step 5: Upload files
        self.upload_files(version_name, files)
        
        # Step 6: Finalize version
        self.finalize_version(version_name)
        
        # Step 7: Create release
        self.create_release(version_name)
        
        # Construct site URL
        if self.channel == "live":
            site_url = f"https://{self.project_id}.web.app"
        else:
            site_url = f"https://{self.project_id}--{self.channel}.web.app"
        
        return site_url


def get_config() -> Tuple[str, str, str, str]:
    """Get configuration from CLI args or environment variables"""
    parser = argparse.ArgumentParser(
        description="Deploy static website to Firebase Hosting using REST API"
    )
    parser.add_argument("--project", help="Firebase project ID")
    parser.add_argument("--token", help="Google OAuth access token")
    parser.add_argument("--dir", default="./public", help="Directory to deploy (default: ./public)")
    parser.add_argument("--channel", default="live", help="Deployment channel (default: live)")
    
    args = parser.parse_args()
    
    project_id = args.project or os.getenv("FIREBASE_PROJECT_ID")
    access_token = args.token or os.getenv("FIREBASE_ACCESS_TOKEN")
    deploy_dir = args.dir or os.getenv("DEPLOY_DIR", "./public")
    channel = args.channel or os.getenv("DEPLOY_CHANNEL", "live")
    
    return project_id, access_token, deploy_dir, channel


def main():
    """Main entry point"""
    try:
        project_id, access_token, deploy_dir, channel = get_config()
        
        # Validate configuration
        if not project_id:
            print("❌ Error: Firebase project ID is required")
            print("Provide via --project=my-project or FIREBASE_PROJECT_ID env var")
            sys.exit(1)
        
        if not access_token:
            print("❌ Error: Access token is required")
            print("Provide via --token=ya29.xxx or FIREBASE_ACCESS_TOKEN env var")
            print("\nTo get an access token:")
            print("1. Go to OAuth 2.0 Playground: https://developers.google.com/oauthplayground/")
            print("2. Select Firebase Hosting API v1beta1 scopes")
            print("3. Authorize and get access token")
            sys.exit(1)
        
        # Create deployer and deploy
        deployer = FirebaseDeployer(project_id, access_token, deploy_dir, channel)
        site_url = deployer.deploy()
        
        # Success!
        print("\n🎉 Deployment successful!")
        print(f"\n🌐 Your site is live at: {site_url}\n")
        
    except Exception as e:
        print(f"\n❌ Deployment failed: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
